[Poh Say Keong] iP

PohSayKeong opened the pull request:
# ZacPro

> “We do not want merely to see beauty. We want something else which can hardly be put into words — to be united with the beauty we see, to pass into it, to receive it into ourselves.” – CS Lewis ([source](https://www.goodreads.com/quotes/8218670-we-do-not-want-merely-to-see-beauty-though-god))

ZacPro tracks your tasks and it's fantastic! Its:
- text-based
- difficult to learn
- ~~slow~~ extremely slow to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
2. Double-click it
3. add tasks
4. Let it manage your life for you 😘

And it is **100 USD**!

Features:
- [ ] Managing tasks
- [ ] Managing deadlines (coming soon!)
- [X] Reminder (not coming)
---

If you are a Java program, you can also use it to practice Java. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

PohSayKeong committed:
Rename, Greet, Exit

PohSayKeong committed:
Echo

PohSayKeong committed:
Add, List

PohSayKeong committed:
Mark as Done

PohSayKeong committed:
ToDos, Events, Deadlines

PohSayKeong committed:
Automated Text UI Testing

PohSayKeong committed:
Handle Errors

PohSayKeong committed:
Delete

PohSayKeong committed:
Add write to file feature

PohSayKeong committed:
Add parse date and time

PohSayKeong committed:
Refactor to make the code more OOP

PohSayKeong committed:
Divide classes into packages

PohSayKeong committed:
Change switch statements to follow code style

PohSayKeong committed:
Automate project builds using Gradle

PohSayKeong committed:
Add JUnit tests for TaskList

PohSayKeong committed:
Add JavaDoc comments

PohSayKeong committed:
Tweak the code to comply with the coding standard

PohSayKeong committed:
Give users a way to find a task by searching for a keyword

PohSayKeong committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/TaskList.java

PohSayKeong committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java

PohSayKeong committed:
Use CheckStyle

PohSayKeong committed:
Merge branch 'branch-A-CheckStyle'

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/task/Deadline.java

PohSayKeong committed:
Add a GUI to the chatbot

pra-navi commented:
looks good but forgot italicising (not sure if the checklist is supposed to be editable by viewers or not 
like the personal touch added to it!

xyT-T commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,103 @@
+            m.find();
+            String taskType = m.group(1);
+            m.find();
+            String taskComplete = m.group(1);
```
Maybe can name it as something like "taskIsComplete" to inform that it represents a boolean value?

xyT-T commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,71 @@
+        String[] userIn = scannerIn.split(" ");
+        String command = userIn[0];
+        switch (command) {
+        case "bye":
```
Good job that you have correct identation for switch! I did it wrongly😢

xyT-T commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,137 @@
+     * @throws TaskNotFoundException If task does not exist.
+     */
+    public void unmarkTask(int taskId) throws TaskNotFoundException {
+        Task unmarkTask = getTask(taskId);
```
Should this object be named as a noun? Perhaps unmarkedTask would be good?

xyT-T commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,137 @@
+     * @throws TaskNotFoundException If task does not exist.
+     */
+    public void deleteTask(int taskId) throws TaskNotFoundException {
+        if (taskId >= taskList.size()) {
```
Another possible exception would be that taskId < 0. Do we need to consider this case here?

xyT-T commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    public static Event initializeFromInput(String input) throws EmptyDescriptionException {
+        try {
+            String processed = input.split("event")[1];
```
Is it possible to name it more clearly?

xyT-T commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,41 @@
+     * @return A new Todo object.
+     * @throws EmptyDescriptionException If todo description is empty.
+     */
+    public static Todo initializeFromInput(String input) throws EmptyDescriptionException {
```
I like how you create an individual method to construct a new ToDo from user input.👍 I believe it makes the interaction between classes clearer and easier to handle.

xyT-T commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Test
+    public void testGetTask() {
```
Maybe you can have a look at the naming convention mentioned in week 3 notes 3.7:

> Java code normally use camelCase for method names e.g., `testStringConversion` but when writing test methods, sometimes another convention is used:
> `unitBeingTested_descriptionOfTestInputs_expectedOutcome`
> e.g., `intDivision_zeroDivisor_exceptionThrown`

but I believe your naming looks good as well!

pra-navi commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+        try {
+            storage.load();
+            tasks = new TaskList(storage.load());
+        } catch (Exception e) {
```
perhaps might be better to label it as 'exception' rather than 'e'?

pra-navi commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+        while (isBotRunning) {
+            try {
+                String scannerIn = scanner.nextLine();
+                isBotRunning = Parser.processCommand(scannerIn, tasks);
```
i like the good choice in variable name here!

pra-navi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,71 @@
+    public static boolean processCommand(String scannerIn, TaskList tasks) throws Exception {
+        String[] userIn = scannerIn.split(" ");
+        String command = userIn[0];
+        switch (command) {
```
i like the clean and neat organisation of code here!

pra-navi commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,103 @@
+        Scanner myReader = new Scanner(file);
+        while (myReader.hasNextLine()) {
+            String data = myReader.nextLine();
+            Matcher m = Pattern.compile("\\[(.*?)]").matcher(data);
```
instead of 'm' it might be clearer to put 'matcher' for future reference in the method

pra-navi reviewed (commented):
LGTM. Just a few nits to fix. I like the organisation and neatness in your code (esp the many different exceptions)!

PohSayKeong committed:
Use Assertions

Currently there may be bugs in the code because the code has resulted in
a program state that violates an assumption about how the code should behave.

Let's add assertions to validate assumptions and check conditions within the code.

PohSayKeong committed:
Improve code quality

Currently there are areas where the code does not follow the guidelines laid out.

Let's try to improve the code quality by also taking into account the feedback given.

PohSayKeong committed:
Merge pull request #1 from PohSayKeong/branch-A-Assertions

Use Assertions

PohSayKeong committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Storage.java

PohSayKeong committed:
Merge pull request #2 from PohSayKeong/branch-A-CodeQuality

Improve code quality

PohSayKeong committed:
Find free times

Currently there is no way for a user to find out when he would be free.

A user may want to find free time to schedule more events.

Let's extend the Duke bot by providing a way for the user to find free times.

PohSayKeong committed:
Fix header comments based on code quality feedback

PohSayKeong committed:
Add a representative screenshot of the product to the docs folder

PohSayKeong committed:
Update README.md

PohSayKeong committed:
Update README.md

PohSayKeong committed:
Add more javadoc comments

PohSayKeong committed:
Fix a bug where the file names of storage are different

PohSayKeong committed:
Update jar build configuration
